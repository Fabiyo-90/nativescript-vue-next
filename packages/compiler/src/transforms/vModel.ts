import {
  DirectiveTransform,
  NodeTypes,
  transformModel as baseTransform,
} from '@vue/compiler-core'
import { V_MODEL } from '@nativescript-vue/compiler'

export const transformModel: DirectiveTransform = (dir, node, context) => {
  const baseResult = baseTransform(dir, node, context)
  baseResult.needRuntime = context.helper(V_MODEL)
  baseResult.props = baseResult.props.filter((p) => {
    if (
      p.key.type === NodeTypes.SIMPLE_EXPRESSION &&
      p.key.content === 'modelValue'
    ) {
      return false
    }
    return true
  })
  return baseResult
}
