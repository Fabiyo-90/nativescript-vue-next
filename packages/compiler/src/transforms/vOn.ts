import {
  transformOn as baseTransform,
  DirectiveTransform,
  createObjectProperty,
  createSimpleExpression,
  createCompoundExpression,
  isStaticExp
} from '@vue/compiler-core'
import { capitalize, makeMap } from '@vue/shared'

// todo: simplify - we only support `once` - no need to use makeMap
const isEventOptionModifier = /*#__PURE__*/ makeMap(`once,capture`)

// todo: remove - no need to loop through modifiers as only `once` is supported
const generateOptionModifiers = (modifiers: string[]) => {
  const eventOptionModifiers = []

  for (let i = 0; i < modifiers.length; i++) {
    const modifier = modifiers[i]

    if (isEventOptionModifier(modifier)) {
      // eventOptionModifiers: modifiers for addEventListener() options, e.g. .once
      eventOptionModifiers.push(modifier)
    }
  }

  return eventOptionModifiers
}

export const transformOn: DirectiveTransform = (dir, node, context) => {
  return baseTransform(dir, node, context, baseResult => {
    const { modifiers } = dir
    if (!modifiers.length) return baseResult

    let { key, value: handlerExp } = baseResult.props[0]
    const eventOptionModifiers = generateOptionModifiers(modifiers)

    if (eventOptionModifiers.length) {
      const modifierPostfix = eventOptionModifiers.map(capitalize).join('')
      key = isStaticExp(key)
        ? createSimpleExpression(`${key.content}${modifierPostfix}`, true)
        : createCompoundExpression([`(`, key, `) + "${modifierPostfix}"`])
    }

    return {
      props: [createObjectProperty(key, handlerExp)]
    }
  })
}
