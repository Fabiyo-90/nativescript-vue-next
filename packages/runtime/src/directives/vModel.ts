import { ObjectDirective, VNode } from '@vue/runtime-core'
import { getViewMeta, NSVModelDescriptor } from '../registry'
import { addEventListener } from '../modules/events'
import { INSVElement } from '../nodes'
import { isArray, invokeArrayFns } from '@vue/shared'

type AssignerFn = (value: any) => void

const getModelAssigner = (vnode: VNode): AssignerFn => {
  const fn = vnode.props!['onUpdate:modelValue']
  return isArray(fn) ? (value) => invokeArrayFns(fn, value) : fn
}

function toNumber(val: string): number | string {
  const n = parseFloat(val)
  return isNaN(n) ? val : n
}

type ModelDirective<T> = ObjectDirective<T & { _assign: AssignerFn }>

export const vModel: ModelDirective<INSVElement> = {
  beforeMount(el, { value, modifiers: { trim, number } }, vnode) {
    el._assign = getModelAssigner(vnode)
    const castToNumber = number
    const { prop, event } = getViewMeta(el.tagName).model as NSVModelDescriptor
    el._assign(value)
    el.setAttribute(prop, value)

    addEventListener(el, event, () => {
      let propValue: unknown = el.getAttribute(prop)
      if (trim && typeof propValue === 'string') {
        propValue = propValue.trim()
      } else if (castToNumber && typeof propValue === 'string') {
        propValue = toNumber(propValue)
      }
      el._assign(propValue)
    })
  },

  beforeUpdate(el, { value, oldValue }, vnode) {
    const { prop } = getViewMeta(el.tagName).model as NSVModelDescriptor
    el._assign = getModelAssigner(vnode)
    if (value === oldValue) {
      return
    }
    el.setAttribute(prop, value)
  },
}
