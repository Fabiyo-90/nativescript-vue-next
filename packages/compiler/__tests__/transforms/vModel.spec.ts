import {
  baseParse as parse,
  transform,
  generate,
  ElementNode,
  ObjectExpression,
  CompilerOptions,
  VNodeCall,
  transformElement
} from '../../src'
import { transformModel } from '../../src/transforms/vModel'

function parseWithVModel(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)

  transform(ast, {
    nodeTransforms: [transformElement],
    directiveTransforms: {
      ...options.directiveTransforms,
      model: transformModel
    },
    ...options
  })

  return ast
}

describe('compiler: transform v-model', () => {
  test('simple expression', () => {
    const root = parseWithVModel('<TextField v-model="model" />')
    const node = root.children[0] as ElementNode
    const props = ((node.codegenNode as VNodeCall).props as ObjectExpression)
      .properties

    expect(props[0]).toMatchObject({
      key: {
        content: 'modelValue',
        isStatic: true
      },
      value: {
        content: 'model',
        isStatic: false
      }
    })

    expect(props[1]).toMatchObject({
      key: {
        content: 'onUpdate:modelValue',
        isStatic: true
      },
      value: {
        children: [
          '$event => (',
          {
            content: 'model',
            isStatic: false
          },
          ' = $event)'
        ]
      }
    })

    expect(generate(root).code).toMatchSnapshot()
  })
})
