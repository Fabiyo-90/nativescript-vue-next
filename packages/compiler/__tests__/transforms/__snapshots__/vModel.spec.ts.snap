// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: transform v-model simple expression 1`] = `
"const _Vue = Vue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { resolveComponent: _resolveComponent, vModel: _vModel, withDirectives: _withDirectives, openBlock: _openBlock, createBlock: _createBlock } = _Vue

    const _component_TextField = _resolveComponent(\\"TextField\\")

    return _withDirectives((_openBlock(), _createBlock(_component_TextField, {
      \\"onUpdate:modelValue\\": $event => (model = $event)
    }, null, 8 /* PROPS */, [\\"onUpdate:modelValue\\"])), [
      [_vModel, model]
    ])
  }
}"
`;
